File: frontend/js/views/lease_money_factor.js
// Car lease payment — depreciation + finance fee from cap cost, residual,
// and money factor, with the equivalent APR, via /calc/lease-payment. Live.

import { api } from '../api.js';
import { applyUiI18n, t } from '../i18n.js';
import { currentViewToken, viewIsCurrent } from '../app.js';
import { showToast } from '../toast.js';
import { debounce } from '../util.js';

const FIELDS = [
    ['cap_cost_usd', 'Capitalized cost ($)', 35000],
    ['residual_value_usd', 'Residual value ($)', 21000],
    ['term_months', 'Term (months)', 36],
    ['money_factor', 'Money factor', 0.00125],
    ['sales_tax_pct', 'Sales tax (%)', 0],
];

const money = (n) => '$' + Number(n).toLocaleString(undefined, { minimumFractionDigits: 2, maximumFractionDigits: 2 });
const money0 = (n) => '$' + Number(n).toLocaleString(undefined, { maximumFractionDigits: 0 });

export async function renderLeasePayment(mount, _appState) {
    const tok = currentViewToken();
    if (!viewIsCurrent(tok)) return;
    mount.innerHTML = `
        <h1 class="view-title"><span data-i18n="view.lmf.h1.title">// CAR LEASE PAYMENT</span></h1>
        <p class="muted small" data-i18n="view.lmf.hint.intro">
            A lease payment is depreciation + finance fee + tax. Depreciation = (cap cost −
            residual) ÷ term — paying for the value lost while you drive. The finance fee =
            (cap cost + residual) × money factor — the interest; the money factor is a disguised
            rate (APR = money factor × 2400). Tax is on the payment in most states. Updates as
            you type.
        </p>
        <div class="lpv-split">
        <div class="chart-panel">
            <h2 data-i18n="view.lmf.h2.inputs">The lease</h2>
            <form id="lmf-form" class="inline-form">
                ${FIELDS.map(([key, label, def]) => `
                    <label><span data-i18n="view.lmf.label.${key}">${label}</span>
                        <input type="number" step="${key === 'money_factor' ? '0.00001' : '0.01'}" min="0" name="${key}" value="${def}" required></label>
                `).join('')}
            </form>
        </div>
        <div id="lmf-result" class="lpv-preview"></div>
        </div>
    `;
    applyUiI18n(mount);

    const form = mount.querySelector('#lmf-form');
    const generate = async () => {
        const fd = new FormData(form);
        const body = {};
        for (const [key] of FIELDS) body[key] = Number(fd.get(key)) || 0;
        try {
            const r = await api.calcLeasePayment(body);
            if (!viewIsCurrent(tok)) return;
            renderResult(mount, r);
        } catch (err) {
            showToast(err.message || t('view.lmf.toast.error'), { level: 'error' });
        }
    };
    form.addEventListener('input', debounce(generate, 250));
    form.addEventListener('submit', (e) => { e.preventDefault(); generate(); });
    generate();
}

function renderResult(mount, r) {
    const el = mount.querySelector('#lmf-result');
    el.innerHTML = `
        <div class="chart-panel">
            <h2 data-i18n="view.lmf.h2.result">The payment</h2>
            <div class="cards">
                <div class="card pos"><div class="label" data-i18n="view.lmf.card.payment">Monthly payment</div>
                    <div class="value pos">${money(r.monthly_payment_usd)}</div></div>
                <div class="card"><div class="label" data-i18n="view.lmf.card.apr">Equivalent APR</div>
                    <div class="value">${Number(r.equivalent_apr_pct).toFixed(2)}%</div></div>
                <div class="card"><div class="label" data-i18n="view.lmf.card.total">Total lease cost</div>
                    <div class="value">${money0(r.total_lease_cost_usd)}</div></div>
            </div>
            <table class="data-table">
                <thead><tr><th data-i18n="view.lmf.col.line">Line</th><th data-i18n="view.lmf.col.amount">Monthly</th></tr></thead>
                <tbody>
                    <tr><td data-i18n="view.lmf.row.dep">Depreciation fee</td><td>${money(r.monthly_depreciation_usd)}</td></tr>
                    <tr><td data-i18n="view.lmf.row.fin">Finance (rent) fee</td><td>${money(r.monthly_finance_usd)}</td></tr>
                    <tr><td data-i18n="view.lmf.row.tax">Tax</td><td>${money(r.monthly_tax_usd)}</td></tr>
                    <tr class="emph"><td data-i18n="view.lmf.row.total">Monthly payment</td><td class="pos">${money(r.monthly_payment_usd)}</td></tr>
                </tbody>
            </table>
        </div>
    `;
    applyUiI18n(el);
}

File: crates/traderview-core/src/lease_money_factor.rs
//! Car lease payment — money factor, depreciation fee, and finance fee.
//!
//! A lease payment has two parts plus tax:
//!
//!   * **Depreciation fee** = (capitalized cost − residual value) / term —
//!     you pay for the value the car loses while you drive it.
//!   * **Finance (rent) fee** = (cap cost + residual) × money factor — the
//!     interest. The money factor is a disguised rate: **APR = money factor
//!     × 2400**.
//!   * **Tax** is charged on the monthly payment in most states.
//!
//! Reports the payment breakdown, the equivalent APR, and the total cost of
//! the lease. Pure compute.

use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Deserialize)]
pub struct LeaseInput {
    /// Capitalized cost — negotiated price + fees − any cap-cost reduction.
    pub cap_cost_usd: f64,
    /// Residual value at lease end.
    pub residual_value_usd: f64,
    pub term_months: f64,
    /// Money factor (e.g. 0.00125 ≈ 3% APR).
    pub money_factor: f64,
    /// Sales tax applied to the monthly payment.
    #[serde(default)]
    pub sales_tax_pct: f64,
}

#[derive(Debug, Clone, Serialize)]
pub struct LeaseResult {
    pub monthly_depreciation_usd: f64,
    pub monthly_finance_usd: f64,
    /// Depreciation + finance, before tax.
    pub monthly_pre_tax_usd: f64,
    pub monthly_tax_usd: f64,
    /// Full monthly payment including tax.
    pub monthly_payment_usd: f64,
    /// Money factor × 2400 — the equivalent APR.
    pub equivalent_apr_pct: f64,
    /// Monthly payment × term.
    pub total_lease_cost_usd: f64,
}

pub fn analyze(i: &LeaseInput) -> LeaseResult {
    let term = i.term_months.max(0.0);
    let depreciation = i.cap_cost_usd - i.residual_value_usd;
    let monthly_dep = if term > 0.0 { depreciation / term } else { 0.0 };
    let monthly_fin = (i.cap_cost_usd + i.residual_value_usd) * i.money_factor;
    let pre_tax = monthly_dep + monthly_fin;
    let tax = pre_tax * i.sales_tax_pct / 100.0;
    let payment = pre_tax + tax;

    LeaseResult {
        monthly_depreciation_usd: monthly_dep,
        monthly_finance_usd: monthly_fin,
        monthly_pre_tax_usd: pre_tax,
        monthly_tax_usd: tax,
        monthly_payment_usd: payment,
        equivalent_apr_pct: i.money_factor * 2400.0,
        total_lease_cost_usd: payment * term,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn base() -> LeaseInput {
        LeaseInput {
            cap_cost_usd: 35_000.0,
            residual_value_usd: 21_000.0,
            term_months: 36.0,
            money_factor: 0.00125,
            sales_tax_pct: 0.0,
        }
    }

    #[test]
    fn monthly_depreciation() {
        // (35k − 21k) / 36 = 388.89.
        let r = analyze(&base());
        assert!((r.monthly_depreciation_usd - 14_000.0 / 36.0).abs() < 1e-6);
    }

    #[test]
    fn finance_fee_uses_sum_times_money_factor() {
        // (35k + 21k) × 0.00125 = 70.
        let r = analyze(&base());
        assert!((r.monthly_finance_usd - 70.0).abs() < 1e-6);
    }

    #[test]
    fn pre_tax_is_dep_plus_finance() {
        let r = analyze(&base());
        assert!((r.monthly_pre_tax_usd - (r.monthly_depreciation_usd + r.monthly_finance_usd)).abs() < 1e-9);
    }

    #[test]
    fn tax_on_payment_and_total() {
        let r = analyze(&LeaseInput { sales_tax_pct: 8.0, ..base() });
        assert!((r.monthly_tax_usd - r.monthly_pre_tax_usd * 0.08).abs() < 1e-9);
        assert!((r.monthly_payment_usd - (r.monthly_pre_tax_usd + r.monthly_tax_usd)).abs() < 1e-9);
    }

    #[test]
    fn equivalent_apr_is_mf_times_2400() {
        // 0.00125 × 2400 = 3%.
        let r = analyze(&base());
        assert!((r.equivalent_apr_pct - 3.0).abs() < 1e-9);
    }

    #[test]
    fn total_cost_is_payment_times_term() {
        let r = analyze(&base());
        assert!((r.total_lease_cost_usd - r.monthly_payment_usd * 36.0).abs() < 1e-6);
    }

    #[test]
    fn higher_money_factor_raises_payment() {
        let low = analyze(&base());
        let high = analyze(&LeaseInput { money_factor: 0.0025, ..base() });
        assert!(high.monthly_payment_usd > low.monthly_payment_usd);
        assert!((high.equivalent_apr_pct - 6.0).abs() < 1e-9);
    }

    #[test]
    fn higher_residual_lowers_payment() {
        // Less depreciation dominates the small finance-fee increase.
        let low_resid = analyze(&LeaseInput { residual_value_usd: 18_000.0, ..base() });
        let high_resid = analyze(&LeaseInput { residual_value_usd: 24_000.0, ..base() });
        assert!(high_resid.monthly_payment_usd < low_resid.monthly_payment_usd);
    }
}
